pkg/ebpf/events_enrich.go

```diff
@@ -0,0 +1,124 @@
+package ebpf
+
+import (
+	gocontext "context"
+
+	"github.com/aquasecurity/tracee/pkg/containers/runtime"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+func (t *Tracee) enrichContainerEvents(ctx gocontext.Context, in <-chan *trace.Event) (chan *trace.Event, chan error) {
+	type enrichResult struct {
+		cgroupId uint64
+		result   runtime.ContainerMetadata
+		err      error
+	}
+
+	queues := make(map[uint64]chan *trace.Event) //map between cgroupId and queues
+	attempted := make(map[uint64]bool)           //map between cgroupId and enrichment attempt
+	enriches := make(chan enrichResult, 10)      //this channel has a small buffer to reduce chance for blocking
+	out := make(chan *trace.Event, 10000)
+	errc := make(chan error, 1)
+	done := make(chan struct{}, 1)
+
+	go func() {
+		defer close(out)
+		defer close(errc)
+		for {
+			select {
+			case <-ctx.Done():
+				done <- struct{}{}
+				return
+			case event := <-in:
+				cgroupId := uint64(event.CgroupID)
+				//if the event is the cgroup_mkdir event we need the cgroupId from it's argument
+				if event.EventID == int(CgroupMkdirEventID) {
+					cgroupId, _ = getEventArgUint64Val(event, "cgroup_id")
+				}
+				//if the event is a cgroup_rmdir we need to clean the attempt entry so the cgroupId can be reused
+				if event.EventID == int(CgroupRmdirEventID) {
+					cgroupId, _ = getEventArgUint64Val(event, "cgroup_id")
+					delete(attempted, cgroupId)
+				}
+				//if non container event and not cgroup_mkdir or the event is already enriched, skip this stage
+				if (event.ContainerID == "" && event.EventID != int(CgroupMkdirEventID)) || event.ContainerImage != "" || attempted[uint64(cgroupId)] {
+					out <- event
+					continue
+				} else {
+					//if the container queue has not been created - create the container queue and invoke the enrich query
+					if _, ok := queues[cgroupId]; !ok {
+						queues[cgroupId] = make(chan *trace.Event, 1000)
+
+						go func() {
+							metadata, err := t.containers.EnrichCgroupInfo(cgroupId)
+							enriches <- enrichResult{cgroupId, metadata, err}
+						}()
+					}
+					queues[cgroupId] <- event
+				}
+			}
+		}
+	}()
+
+	go func() {
+		for enrich := range enriches {
+			cgroupId := enrich.cgroupId
+			//mark the query as finished, it should not be attempted again and no new events should enter it's queue
+			attempted[cgroupId] = true
```

Comment:
If you do this here.. and an event for this cgroupid arrives in the other routine, it will be sent to the out channel BEFORE the buffered events, making them out of order, no ?

Reply:
Ok so we need to block events from entering the queue when we are processing it AND, at the same time, we can't allow that event to be sent to "out" channel just yet.